### PR TITLE
Add app stats fetch

### PR DESCRIPTION
**Pull request by kratos7**
Query slaves to get app stats

### CODE REVIEW

File: src/main/python/hydra/lib/hydrabase.py [hunk 1]
```diff
@@ -666,6 +671,66 @@ def get_mesos_slave_ips_attr(self, attr_type, attr_value):
         """
         return self.__mesos.get_slave_ips_from_attribute(attr_type, attr_value)
 
+    def get_all_mesos_slave_attr(self):
```
Comment (tahir24434) [inline src/main/python/hydra/lib/hydrabase.py:761]: Is it necessary to put dictionary datastructure info to list. Can't we just do something like below

for idx, info in self.mesos_cluster.items():
  self.get_mesos_slave_ips_attr(info["cat"], info["match"])

Lesser number of functions (data structures) will save time needed to understand the code.

File: src/main/python/hydra/lib/hydrabase.py [hunk 2]
```diff
@@ -666,6 +671,66 @@ def get_mesos_slave_ips_attr(self, attr_type, attr_value):
...
+        """
+        app_stats_list = {}
+        task_stats = {"timestamp": 0,
+                      "mem_rss_bytes": 0,
+                      "cpu_system_time_secs": 0,
+                      "cpu_user_time_secs": 0,
+                      }
+        assert(name in self.app_id_list)
+        if not self.all_mesos_slave_iplist:
+            self.all_mesos_slave_iplist = self.get_all_mesos_slave_iplist()
```
Comment (tahir24434) [inline src/main/python/hydra/lib/hydrabase.py:803]: get_all_mesos_slave_iplist function raises an exception. What about catching that exception here?

File: src/main/python/hydra/lib/mmapi.py [hunk 3]
```diff
@@ -179,6 +179,14 @@ def get_health(self):
             return True
         return False
 
+    def get_slave_stats(self, ip, port=5051):
```
Comment (tahir24434) [inline src/main/python/hydra/lib/mmapi.py:185]: Please add function comments.

File: src/main/python/hydra/lib/mmapi.py [hunk 4]
```diff
@@ -179,6 +179,14 @@ def get_health(self):
...
+        addr = "http://%s:%d" % (ip, port)
+        # returns a list of app tasks
+        r = requests.get(addr + '/monitor/statistics.json')
+        assert(r.status_code == 200)
```
Comment (tahir24434) [inline src/main/python/hydra/lib/mmapi.py:195]: Does request.get returns 200 in case of success?

File: src/main/python/hydra/lib/hydrabase.py [hunk 5]
```diff
@@ -666,6 +671,66 @@ def get_mesos_slave_ips_attr(self, attr_type, attr_value):
...
+        for slave_ip in self.all_mesos_slave_iplist:
+            # list of all app and its tasks
+            stats = self.__mesos.get_slave_stats(slave_ip)
+            assert(stats)
+            for app_task in stats:
```
Comment (tahir24434) [inline src/main/python/hydra/lib/hydrabase.py:808]: This comment can be ignored.
Perhaps, we can do following
desired_app_task = (app_task for app_task in stats if app_task["source"] == "name").next()

File: src/main/python/hydra/lib/mmapi.py (hunk 3 above)
Comment (kratos7) [reply src/main/python/hydra/lib/mmapi.py:185]: ok

File: src/main/python/hydra/lib/hydrabase.py (hunk 2 above)
Comment (kratos7) [reply src/main/python/hydra/lib/hydrabase.py:803]: why ? Thats a valid exception.

File: src/main/python/hydra/lib/mmapi.py (hunk 4 above)
Comment (kratos7) [reply src/main/python/hydra/lib/mmapi.py:195]: yes. its http success status code.

File: src/main/python/hydra/lib/hydrabase.py (hunk 5 above)
Comment (kratos7) [reply src/main/python/hydra/lib/hydrabase.py:808]: I think explicit code is more readable and easier to understand. The one liner is confusing, requires more time to read, and achieves no added benefit IMO.

File: src/main/python/hydra/lib/hydrabase.py (hunk 1 above)
Comment (kratos7) [reply src/main/python/hydra/lib/hydrabase.py:761]: not sure I understand the comment, what data structure to info list are we talking about here ?

File: src/main/python/hydra/lib/hydrabase.py (hunk 2 above)
Comment (tahir24434) [reply src/main/python/hydra/lib/hydrabase.py:803]: Yes, thats a valid exception. What I was trying to say was that when get_all_mesos_slave_iplist() will raise that exception, our test case will stop immediately. If we surround the function call in try catch statement, we can handle the exception in whatever way we want in Except part.

File: src/main/python/hydra/lib/hydrabase.py (hunk 1 above)
Comment (tahir24434) [reply src/main/python/hydra/lib/hydrabase.py:761]: So, we have a dictionary datastructure named 'info'. We are putting info dict values to another list named attr_list. And then we are using that list to do whatever we want to do.
The thing which I was trying to ask is that can we avoid the intermediate list attr_list DS? Can't we directly use the info dict.

Comment (kratos7) [reply src/main/python/hydra/lib/hydrabase.py:761]: Ok Now i get it. the comment is not clear :)
Yes I can just use self.mesos_cluster in place, but I foresee the need of a function to just get the attr list of all the registered slaves.
Adding an APi to get that doesnt harm in my opinion. We might need to use it later as well.

File: src/main/python/hydra/lib/hydrabase.py (hunk 2 above)
Comment (kratos7) [reply src/main/python/hydra/lib/hydrabase.py:803]: Yes thats what the intended behavior is, the user expects to get a valid stats response after calling this function, if its unable to find the ip of a registered slave, that program should assert and exit.

File: src/main/python/hydra/lib/mmapi.py (hunk 3 above)
Comment (kratos7) [reply src/main/python/hydra/lib/mmapi.py:185]: done

File: src/main/python/hydra/lib/hydrabase.py (hunk 2 above)
Comment (kratos7) [reply src/main/python/hydra/lib/hydrabase.py:803]: And any decision can be made by the user e-g kraken, that will actually poll the stats using these functions.

**Comment by tahir24434** [comment]
Code looks ok to me. Conflicts need to be resolved though.

**Comment by tahir24434** [comment]
Code looks fine.

**Pull request merged**